### PR TITLE
Feat: Discount migration

### PR DESCRIPTION
**Pull request by pKorsholm**
**What** 
- Update discount migration command for upgrading the datamodel using the command `medusa migrations datamodel run` 

How
- Using a querybuilder in the migration command, which in batches updates the datamodel. The querybuilder command is modelled on the sql queries in the notion doc attached under the up section.

**Why**
- To automate the migration of data when upgrading the medusa version and running migrations
- Not migrating the data would cause unexpected behaviour for the customer since discounts would instead be valid for all products and not just a select few (due to the structure of conditions)

**Testing** 
- For now I have tested the migration using database queries, however it should be re-tested in a real store environment to ensure that discounts are applied correctly

- Test data here: https://medusajs.notion.site/Discount-migration-script-5feb098b1d4b422c829539c17c127134

### CODE REVIEW

**Review by srindom — CHANGES_REQUESTED** [review]
Looks good! Would maybe suggest that we move this into a separate script that can be executed independently of the migration; primarily to avoid potentially having a very long running migration.

We could add a small check along the lines of:

```
SELECT id from discount_rule_products limit 1;
```

And then if nothing is returned we can safely drop `discount_rule_products`.

wdyt? :)

**Comment by pKorsholm** [comment]
I agree we should add a check to see that the rows have been correctly inserted. I think we should run these queries: 

```sql
 select COUNT(*), COUNT(DISTINCT(discount_rule_id)) from discount_rule_products
 select COUNT(*), COUNT(DISTINCT(condition_id)) from discount_condition_product
```

The results of these queries can be checked that they are identical which, if they are, should mean that all discount rules have been inserted and grouped into conditions like they were grouped to `discount_rule`s before. This doesn't validate that the products are connected to the correct `discount_rule`, however the check that the grouping counts matches I think is okay for now. We could run an additional query to match each of the entries in `discount_condition_product` to a `discount_rule`. 

I think in terms of moving this out to a separate script to be run individually, I discussed possible solutions with Zak and Oli seperately, and my general takeaway was that this might as well run as part of the migration script, though I'm not opposed to creating a separate file with `up` and `down` methods and invoke the migration functions in the corresponding functions in the migrations file. 

The reasoning behind running the migration as part of the script is to avoid any invalid state in the core application since in the time between running the migration and running the sql update the discounts which is set up with `valid_for` will no longer apply due to changes in the core logic of the application. Along with this, I have tested the migration script with 50000 entries in the `discount_rule_products` and it delayed the migration with less than a second in total. I think that even in a worst case scenario this will not affect migrations significantly with the shops using medusa now. With regard to future shops this change will not affect setting up shops since the `discount_rule_products` table will be empty when the migration is applied. 

I think what we should do moving forward is add a couple of guards to check that the table exists as well as a guard to check that everything is migrated to the new tables before removing the old table in the `up` function.

Considering the `down` code we should do something similar to check that all rows have been moved. We should be able to use the same comparison as with the `up` code to ensure that we have all rows migrated before proceeding.

What do you think @srindom?

**Review by srindom — COMMENTED** [review]
I am a bit unsure if I think it is a good idea to add the command to the CLI before we settle on a way to go forward with this in the future. 

Do you think we could maybe move this into `packages/medusa/src/scripts/[name-of-migration].ts` and then as a starting point have the user manually go in and run `node_modules/@medusajs/medusa/dist/scripts/[name-of-migration].ts`?

Alternatively, we could have the raw SQL script in our docs and ask people to copy it into something like pgAdmin. 

Alternatively alternatively, it may make sense to add a `medusa scripts [name of script]` command that could be used to run ad-hoc scripts, but I think we should all sync on this if we decide to go down that route.

I think it is a bit tricky to figure out what the right approach is and before we are confident we have a good path I don't think we should make changes to the CLI. Wdyt? :)

File: packages/medusa/src/scripts/discount-rule-migration.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      .from(DiscountRule, "dr")
+      .getCount()
+
+    // INSERT INTO discount_condition(id, operator, type, discount_rule_id)
+    // (SELECT ROW_NUMBER() OVER(ORDER BY sq.d_id), sq.op, sq.tp, sq.d_id FROM
+    // 	(SELECT DISTINCT
+    // 			'in'::discount_condition_operator_enum op,
+    // 			'products'::discount_condition_type_enum tp,
+    // 			discount_rule_products.discount_rule_id d_id
+    // 	FROM discount_rule_products) sq)
```
Comment (olivermrbl) [inline packages/medusa/src/scripts/discount-rule-migration.ts:41]: issue: Remove

File: packages/medusa/src/scripts/discount-rule-migration.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    let offset = 0
+    while (offset < discountRuleCount) {
+      const test = await manager
```
Comment (olivermrbl) [inline packages/medusa/src/scripts/discount-rule-migration.ts:45]: nit: Fix naming  
```suggestion
      const discountRules = await manager
```

Change to be reflected on line 64

File: packages/medusa/src/scripts/discount-rule-migration.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      offset += BATCH_SIZE
+    }
+
+    const discountRuleProductCount = await manager.query(
+      "SELECT COUNT(*) FROM discount_rule_products" // count old "valid_for"
+    )
+
+    // INSERT INTO discount_condition_product(condition_id, product_id)
+    // (SELECT cond.id, drp.product_id FROM discount_rule_products as drp
+    // LEFT JOIN discount_condition as cond ON cond.discount_rule_id = drp.discount_rule_id)
```
Comment (olivermrbl) [inline packages/medusa/src/scripts/discount-rule-migration.ts:83]: issue: Remove

**Review by olivermrbl — APPROVED** [review]
LGTM! A couple of nits before merging 👍 

cc @srindom

File: packages/medusa/src/scripts/discount-rule-migration.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    let offset = 0
+    while (offset < discountRuleCount) {
+      const discountRules = await manager
+        .createQueryBuilder()
+        .from(DiscountRule, "dr")
+        .select("dr.id", "dr_id")
+        .innerJoin(
+          "discount_rule_product",
+          "drp",
+          "dr_id = drp.discount_rule_id"
```
Comment (olivermrbl) [inline packages/medusa/src/scripts/discount-rule-migration.ts:44]: issue: The alias is strangely not being picked up here. We can just remove it and use the id directly
```suggestion
          "dr.id = drp.discount_rule_id"
```

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Have added a couple of suggestions to issues discovered after having run it against a Medusa project

File: packages/medusa/src/scripts/discount-rule-migration.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      await manager
+        .createQueryBuilder()
+        .insert()
+        .into(DiscountCondition)
+        .values(
+          discountRules.map((dr) =>
+            Object.assign(new DiscountCondition(), {
+              type: DiscountConditionType.PRODUCTS,
+              operator: DiscountConditionOperator.IN,
+              discount_rule_id: dr.dr_id,
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa/src/scripts/discount-rule-migration.ts:60]: Apologies for this one. Should be dr_id. Have made the change 👍

**Review by olivermrbl — APPROVED** [review]
LGTM!

**Review by srindom — APPROVED** [review]

**Pull request merged**